keps/sig-apps/3329-retriable-and-non-retriable-failures/README.md

```diff
@@ -0,0 +1,1267 @@
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+### Reason in DeleteOptions
+
+The `DeleteOptions` is extended by the reason field to convey the reason of
+deletion.
+
+```golang
+// DeleteOptions may be provided when deleting an API object.
+type DeleteOptions struct {
+  ...
+	// A brief CamelCase message indicating the reason why the object is deleted (example: "Preemption").
+	// + optional
+	Reason *string `json:"reason,omitempty" protobuf:"bytes,6,rep,name=reason"`
+}
+```
+
+The `reason` field will be used by kubernetes components (such as:
+kube-scheduler, kubelet pod-gc) when they send the delete operation. During
+the implementation process, we are going to review the places were the
+`DeleteOptions` is used to delete pods and add the `reason` field based on the
+invocation context.
+
+Notably, the `reason` field might be useful to be supplied when deleting other
+types of kubernetes resources, but supplying the field in other contexts is
+beyond this KEP.
+
+In the case of Pods, kube-apiserver will store the received `reason` in the
+already existing `reason` field in the `PodStatus`:
+```
+// PodStatus represents information about the status of a pod. Status may trail the actual
+// state of a system.
+type PodStatus struct {
+  ...
+	// A brief CamelCase message indicating details about why the pod is in this state. e.g. 'Evicted'
+	// +optional
+	Reason string
+  ...
+}
+```
+
+### JobSpec API
+
+We extend the Job API in order to allow to apply different actions depending
+on the conditions associated with the pod failure.
+
+```golang
+// BackoffPolicyAction specifies how a Pod failure is handled.
+type BackoffPolicyAction string
+
+const (
+
+	// This is an action which might be taken on pod failure - mark the
+	// pod's job as Failed.
+	TerminateBackoffAction BackoffPolicyAction = "Terminate"
+
+	// This is an action which might be taken on pod failure - the pod will be
+	// restarted and the counter for .backoffLimit will not be incremented.
+	IgnoreBackoffAction BackoffPolicyAction = "Ignore"
+
+)
+
+// BackoffPolicyRule specifies a backoff policy rule, which comprises an action
+// taken on a pod failure and the condition which needs to be satisfied to take
+// the action.
+type BackoffPolicyRule struct {
+
+	// Specifies the action taken on a pod failure when the condition is satisfied.
+	Action BackoffPolicyAction
+
+	// Restricts the check for exit codes to only apply to the container with
+	// the specified names.
+	// +optional
+	ContainerName *string
+
+	// Specifies the rule condition by the comma-separated list of exit codes or
+	// ranges of exit codes. The condition is satisfied if the actual container's
```

Comment:
You can say things like:

When this field is in use, field X needs to be empty.